Add Reference Popovers

archdragon opened the pull request:
# Add reference popovers

![image](https://camo.githubusercontent.com/80d0c59b303f9cde6bf7b9e63bb41e63ee4f891a/687474703a2f2f6b726f77696e736b692e636f6d2f65785f646f632f706f706f766572732e676966)


|<img src="https://user-images.githubusercontent.com/1810732/59644952-9349de80-916f-11e9-9e43-9b5dba367f6a.png" width="400" />|<img src="https://user-images.githubusercontent.com/1810732/59644974-ae1c5300-916f-11e9-9c97-2d00e16a1119.png" width="400" />|
|---|---|
|<img src="https://user-images.githubusercontent.com/1810732/59639135-94bcdc00-915a-11e9-9613-e6ba2bb145dc.png" width="400" />||

## Summary
Adds popovers discussed in: https://github.com/elixir-lang/ex_doc/issues/971 - tooltips with extra info that appear after hovering over a link, Wikipedia style. :)

## Demo

### [Demo Link (After fixes)](http://krowinski.com/ex_doc/popovers_with_fixes/elixir/Kernel.html)
### ~~[ [Old Demo Link](http://krowinski.com/ex_doc/popovers_done/elixir/) ]~~

## 1. What triggers a reference popover

Reference popover will appear after you hover over a:

* ➡️ Link to a function/callback - function signature and summary will be displayed
  ![image](https://user-images.githubusercontent.com/1810732/59642537-ba031780-9165-11e9-9ef5-7133c7fcaed6.png)


* ➡️Module - module name and summary will be displayed
  ![image](https://user-images.githubusercontent.com/1810732/59642576-ddc65d80-9165-11e9-94d2-905cc49dfde1.png)


* ➡️Types
  * ➡️If defined in the module - name and signature will be displayed
    ![image](https://user-images.githubusercontent.com/1810732/59642759-a7d5a900-9166-11e9-94ed-317d247a5978.png)


  * ➡️If defined on the typespecs page
    * ➡️For built-in types - popover will extract detailed info  
      ![image](https://user-images.githubusercontent.com/1810732/59642657-30a01500-9166-11e9-9c4b-fffff4ac63d1.png)

    * ➡️For basic types and literals - basic hint will be displayed
      ![image](https://user-images.githubusercontent.com/1810732/59642689-5d542c80-9166-11e9-93f9-600693f9000d.png)


## 2. Handling types

I'm able to show info about the built-in types, because they are displayed as a neat table - so it's easy to find the requested type and the corresponding definition.

For basic types and literals this is not really possible - their definitions are unstructured and displayed as one big `<code>` block.

![image](https://user-images.githubusercontent.com/1810732/59644247-73fd8200-916c-11e9-8e38-3cee69d5598c.png)


Maybe we could display them in a table instead? Then I'll be able to upgrade popovers descriptions. :)

## 3. Handling external links

External links - communication between different domains/projects - is fully supported.

For example if https://hexdocs.pm/elixir contains the popover code, links to `https://hexdocs.pm/elixir(...)` will trigger appropriate popovers.

## 4. Security

Incoming data is always displayed as plain text to guard against potential malicious payload.

## 5. Switching `on` and `off`

Tooltips can be disabled with the link at the bottom of the page. Same as with Night Mode, user's choice is stored in localStorage.

![image](https://user-images.githubusercontent.com/1810732/59644765-a9a36a80-916e-11e9-88ef-653ac332b4e8.png)


## 6. Delay

Popovers will appear after a short delay. This way they won't get trigger on accident (ie. while scrilling the page or just moving cursor around).

## 7. How it's implemented

Script is based on José's suggestion to use an iframe to display the popover content.

I've started with a basic iframe displaying a simplified version of the requested page. That worked great for smaller, simpler functions. Longer function names/signatures were overflowing and iframe resizing was needed. In addition iframe content was sometimes "jumping" even after the page was fully loaded (I suspected some other JS scripts adding styles and causing content to shift).

At this point I was already sending messages (containing size measurements and load progress info) from the iframe to the parent page - I've decided to try extracting tooltip title and description using JS and sending them to the parent via postMessage.

This is how it works right now:
1. User hovers over a link
2. A param '?hint=true' is added to the url.
3. Url is loaded inside the iframe
4. When '?hint=true' is detected Javascript tries to find name and description of the module/function/type requested in the url.
5. If it's successful it send them to the parent via window.postMessage
6. Parent receives the postMessage and displays the popover

wojtekmach commented on assets/js/tooltips/tooltips.js:
```diff
@@ -0,0 +1,321 @@
+  return false
+}
+
+// If tooltips are disabled switches them on. If they are enabled switches them on.
```
```suggestion
// If tooltips are disabled switches them on. If they are enabled switches them off.
```

wojtekmach commented on assets/js/tooltips/tooltips.js:
```diff
@@ -0,0 +1,321 @@
+
+/**
+ * Generates an id that will be included, as a param, in the iFrame URL.
+ * Message that comes back from the iFrame will send back this id - this will help to avoid reace conditions.
```
```suggestion
 * Message that comes back from the iFrame will send back this id - this will help to avoid race conditions.
```

wojtekmach commented:
Just wanted to say this is amazing.

Couple minor points:

<img width="536" alt="image" src="https://user-images.githubusercontent.com/76071/59675116-ea2dd300-91c4-11e9-907b-ba5178ce6c4f.png">

(http://krowinski.com/ex_doc/popovers_done/elixir/GenServer.html#module-example)

This is unfortunate result of security considerations above, we rightfully so don't interpret any markup and so we can't do indentation, new lines etc. What if for each spec DOM element we store a data attribute with spec in plaintext (with spaces and newlines) and use that to render in popover? This would obviously increase the size of the docs unfortunately.

What do you think about making it possible to hover over content of the popover? (Currently it's not possible.) This way we'd be able to copy-paste stuff, e.g. complicated specs. Right now what we can do is click on the link and copy-paste the contents from the page we land on.

Both issues are minor and if anything we should work on them in future PRs.

wojtekmach commented:
Just wanted to say this is amazing.

Couple minor points:

<img width="536" alt="image" src="https://user-images.githubusercontent.com/76071/59675116-ea2dd300-91c4-11e9-907b-ba5178ce6c4f.png">

(http://krowinski.com/ex_doc/popovers_done/elixir/GenServer.html#module-example)

This is unfortunate result of security considerations above, we rightfully so don't interpret any markup and so we can't do indentation, new lines etc. What if for each spec DOM element we store a data attribute with spec in plaintext (with spaces and newlines) and use that to render in popover? This would obviously increase the size of the docs unfortunately.

What do you think about making it possible to hover over content of the popover? (Currently it's not possible.) This way we'd be able to copy-paste stuff, e.g. complicated specs. Right now what we can do is click on the link and copy-paste the contents from the page we land on.

Both issues are minor and if anything we should work on them in future PRs.

josevalim commented on assets/test/tooltips/hints-extraction.spec.js:
```diff
@@ -0,0 +1,137 @@
+      expect(extractModuleHint(modulePageObject).kind).to.eql('module')
+    })
+
+    it('extracts plain text, without html tags', () => {
```
I think this can be removed? :D

josevalim commented on assets/js/tooltips/tooltips.js:
```diff
@@ -0,0 +1,321 @@
+}
+// Tooltip will appear only if the mouse cursor stays on the link for at least 150ms.
+// This way tooltips will not appear if we are scrooling the page or just moving the cursor around.
+const hoverDelayTime = 150
```
Because there is a loading time, I think we should be fine with 100 here. :D (bike shedding)

archdragon commented:
Thanks @wojtekmach , @josevalim !

> This is unfortunate result of security considerations above, we rightfully so don't interpret any markup and so we can't do indentation, new lines etc. What if for each spec DOM element we store a data attribute with spec in plaintext (with spaces and newlines) and use that to render in popover? This would obviously increase the size of the docs unfortunately.

Actually adding newline support should be super easy - I'm not removing inner whitespaces, just html tags. A `white-space: pre` CSS should fix this!

> What do you think about making it possible to hover over content of the popover? (Currently it's not possible.) This way we'd be able to copy-paste stuff, e.g. complicated specs. Right now what we can do is click on the link and copy-paste the contents from the page we land on.
>
> Both issues are minor and if anything we should work on them in future PRs.

👍 I'll need to experiment with it a little bit. This I would definitely move to another PR, since it requires a bit of tinkering to get right :)

> 1. [discussion] Should we show popovers for the current page itself? For example, Kernel shows popovers for Kernel.

Personally I would vote for disabling them.

At first I thought that, if I'm a newbie, a quick summary of the current module can still be useful. But after checking again I can see that current module is referenced mostly at the very top of the page, in the first description block - and popover does not make much sense there.

> 2. [bug] Currently, every time a popover is shown, we add a new entry to the browser's history (or at least we do so in Firefox)

On it!

> 3. [discussion] the type stuff is really nice but part of me says we should keep it simple and simply show "Basic / built-in type" with no description of it. I.e. we just look if the link is to "typespecs.html" and if so, we show a short description.

I wanted to see what I can do with types, but I feel that simpler tip might indeed be better.
a) Removing built-in types extraction will greatly simplify the code.
b) Typespecs tips will look a bit more consistent

> 4. [discussion] One of the features we want to have from popovers is to show popovers across projects. Therefore I wonder if the popover itself should include some information about the project the popover is from, possibly in small font, either in the header or in the footer. Such as "Elixir v1.9.2". I believe this information can be found in the meta tag.

👍I'll experiment with the styles/positioning and prepare a demo.

cybrox commented:
> 1. [discussion] Should we show popovers for the current page itself? For example, Kernel shows popovers for Kernel.

I would also vote for disabling them, they get excessive quite fast in modules that reference lots of their own types/functions.

> 4. [discussion] One of the features we want to have from popovers is to show popovers across projects. Therefore I wonder if the popover itself should include some information about the project the popover is from, possibly in small font, either in the header or in the footer. Such as "Elixir v1.9.2". I believe this information can be found in the meta tag.

I like this.

fertapric commented on assets/js/tooltips/hints-page.js:
```diff
@@ -0,0 +1,128 @@
+  pathnameEnd: '/typespecs.html',
+  categories: [
+    { name: 'basicType', description: 'Basic type', hash: '#basic-types', detailsAvailable: false },
+    { name: 'literal', description: 'Literal', hash: '#literals', detailsAvailable: false },
```
I think ExDoc doesn't link typespec literals, do you have any examples of this tooltip showing? 🙂

archdragon replied:
> I think ExDoc doesn't link typespec literals, do you have any examples of this tooltip showing? 🙂

Ah, that makes sense!  I've not seen them anywhere, but added this category just in case, but I guess I can safely remove it. :)

josevalim replied:
Given that the plan is to remove the typespec handling anyway, we don't have to worry about this particular bit either. :D

archdragon commented:
I've just uploaded the updates and fixes :)

➡️ New demo: http://krowinski.com/ex_doc/popovers_with_fixes/elixir/Kernel.html

## Changelog:

### 1. Removed type extraction for `built-in` and `basic` types.
Now only a simple hint is being shown.
![image](https://user-images.githubusercontent.com/1810732/60687748-d5ce2380-9eb0-11e9-9c06-6fd6bdac4553.png)

### 2. Links to the current module page will not trigger the popover.
![image](https://user-images.githubusercontent.com/1810732/60687841-6ad11c80-9eb1-11e9-8ccf-d2a43c09f8b1.png)
^ No popover when hovering over `Kernel`

### 3. Added version info
Displayed in the top right corner.

![image](https://user-images.githubusercontent.com/1810732/60687882-cb605980-9eb1-11e9-8980-94546e813f3a.png)

### 4. Fixed multiple history entries
Tooltips no longer add new history entries.

### 5. Adjusted delay
Basic delay between hover and tooltip showing up changed from 150 do 100ms.

### 6. Fixed white space handling in long signatures

Formatting is now preserved:

![image](https://user-images.githubusercontent.com/1810732/60688032-ec757a00-9eb2-11e9-983e-602bc65cc226.png)

### 7. Minor style adjustments and fixes :)

josevalim commented:
Absolutely beautiful! :heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

josevalim commented:
I think I have found two hopefully tiny bugs:

1. Pop-over does not work when we are linking to a lesser arity. To reproduce it:

  a. Go to this page: http://krowinski.com/ex_doc/popovers_with_fixes/elixir/Access.html
  b. Search for "See the functions key/1, key!/1, elem/1, and all/0 for some of the available accessors."
  c. Mouse over "key/1", it will show the module documentation and not the documentation for key/1. Note that `key/1` exists but it is simply a reference to `key/2`. Perhaps the current code does not work with references like that?

2. References to URLs with weird characters do not work. To reproduce:

  a. Go here: http://krowinski.com/ex_doc/popovers_with_fixes/elixir/Enum.html#sort_by/3
  b. Mouse over `Kernel.<=/2` and it will show the docs for Kernel. I believe this is caused by the escaping in the URL :)

cybrox commented:
For `2.`, I suspect this is because of [this bit](https://github.com/elixir-lang/ex_doc/pull/1025/files#diff-45395fb91680c6517b985dd2b880a4beR80). Back when I solved an issue with encoded/unencoded anchors, we decided to support both, however, the unencoded version is stored in a `span` element inside `.detail` which will not match `$hash.detail`, so the hash would need to be urlencoded to always work with the section id.

```
<section class="detail" id="%3C=/2">
    <span id="<=/2"></span>
```

archdragon commented:
> I think I have found two, hopefully tiny, bugs

> For 2., I suspect this is because of this bit. Back when I solved an issue with encoded/unencoded anchors, we decided to support both, however, the unencoded version is stored in a span element inside .detail 

Thanks for the info - should be a quick fix! :)

archdragon commented:
I've fixed both reported issues :)

➡️ New demo: [Demo Link](http://krowinski.com/ex_doc/popovers_with_fixes2/elixir/Kernel.html) (new one in next post)

1. Fixes the arity links

> Note that key/1 exists but it is simply a reference to key/2. Perhaps the current code does not work with references like that?

Exactly - I missed the fact that we create separate anchors for that. 

Now it's fixed. :)

> Maybe we could adjust these generated span's with a alias class, so that they are identifiable 

Sounds good, but that was not necessary - they are always direct descendants of the `.description` and that makes them easy to find. :)

![image](https://user-images.githubusercontent.com/1810732/61010943-8f793880-a378-11e9-9d99-efc1f5551b20.png)
![image](https://user-images.githubusercontent.com/1810732/61011000-d830f180-a378-11e9-823e-9e79972e741f.png)

2. Fixes encoding/links to weird characters

![image](https://user-images.githubusercontent.com/1810732/61011452-eda71b00-a37a-11e9-8017-ccf05f3121e1.png)

josevalim commented:
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

michallepicki commented:
This looks cool! I have one idea though: wouldn't it be more useful if the displayed version was the `since` metadata instead of the current version? The current version is always displayed on the page in the same spot. If `since` is not present, we could leave this space empty

archdragon commented:
> This looks cool! I have one idea though: wouldn't it be more useful if the displayed version was the since metadata instead of the current version? The current version is always displayed on the page in the same spot. If since is not present, we could leave this space empty

Thanks @michallepicki 

Version info was added because tooltips work across projects (provided that both projects use the latest ex_doc version).

I think `since` would work great when the project references itself, but I'm not sure how to handle that cross-project without making the info confusing.

josevalim commented:
@archdragon one option is: if since is available, we show "since Elixir v1.8.0", otherwise we show "seen on Elixir v1.8.0" or similar. WDYT?

michallepicki commented:
@archdragon cool, where can I see it in action? Here hovering over ExUnit.DocTest doesn't do anything for me: https://hexdocs.pm/elixir/master/writing-documentation.html#doctests , similarily here Logger stuff: https://hexdocs.pm/elixir/master/library-guidelines.html#avoid-macros

michallepicki commented:
@archdragon I just re-read your response and realized - both need to use latest ex_doc, understood

wojtekmach commented:
> ExUnit.DocTest doesn't do anything for me: https://hexdocs.pm/elixir/master/writing-documentation.html#doctests

the reason is ExUnit.DocTest link goes to https://hexdocs.pm/ex_unit/ExUnit.DocTest.html (which was generated with older ex_doc version). You can see it in action here: https://hexdocs.pm/elixir/master/Kernel.html as long as hover over something that stays within Elixir docs (and not e.g. Mix)

archdragon commented:
> @archdragon one option is: if since is available, we show "since Elixir v1.8.0", otherwise we show "seen on Elixir v1.8.0" or similar. WDYT?

Makes sense!

I'll need to spend just a bit more time on it to make sure it looks good. With `since`/`seen on` version name will occupy more space and I might need to adjust CSS rules a bit to account for that.

josevalim commented:
@archdragon one option is to do this: "Elixir v1.8.0" (without since) and "Elixir v1.3.0+" (with since).
